Committing UXUI-209 for Credit Card Number Input

manifestinteractive opened the pull request:
## Credit Card Number Input

New component for https://holidayextras.jira.com/browse/UXUI-209

Passes `npm test` and `npm run lint`

#### Attributes

`accepted` Array - Cards to accept ( 'amex', 'discover', 'mastercard' &amp; 'visa' )
`formatInput` Boolean - ( default false ) Whether to Format Input for easier reading
`showImages` Boolean - ( default false ) Whether to show Credit Card Images
`defaultValue` String - ( default empty ) Initial Value of Credit Card Number
`name` String - ( default 'credit-card-number' ) Name of Input
`id` String - ( default 'credit-card-number' ) ID for Input

## Example

```js
var cc = {
  images: true,
  format: true,
  number: '4485148248332663',
  label: 'Credit Card: ',
  accepted: ['visa','mastercard']
};

var example = (
  <UIToolkit.CreditCardNumber
    showImages={cc.images}
    formatInput={cc.format}
    defaultValue={cc.number}
    label={cc.label}
    accepted={cc.accepted} />
);

React.render(example, mountNode);
```

## Screenshot

![screenshot](http://i.imgur.com/58HNmHP.png)

---
#### Developer Definition of Done/Quality Checklist (for PR author to complete BEFORE code review):
- [x] I have checked the [contributing document](../blob/master/CONTRIBUTING.md) and I'm happy for this to be reviewed.

#### Software Engineer or Developer review:
- [x] I’ve witnessed the work behaving as expected (this could be on the authors machine or screencast).
- [x] I’ve checked for coding anti-patterns.
- [x] I’ve checked for appropriate test coverage.
- [x] I’ve run all the tests locally and they pass.

#### Software Engineer or project guru review:
- [ ] I’ve witnessed the work behaving as expected (this could be on the authors machine or screencast).
- [ ] I’ve checked for coding anti-patterns.
- [ ] I’ve checked for appropriate test coverage.
- [ ] I’ve run all the tests locally and they pass.

stuartstobie commented:
Looking at this now

stuartstobie commented on src/components/credit-card-number/code/templates/creditCardNumberTemplate.jsx:
```diff
@@ -0,0 +1,51 @@
+  }
+
+  if(component.props.showImages)
+  {
```
Keep the { on the same lie as the if, as you have done in the previous if statement. Same applies below.

cameronviner commented:
Taking DEV

cameronviner commented:
Conflicts on this PR, will need to merge in master

cameronviner commented on docs/assets/style.css:
```diff
@@ -130,3 +130,41 @@ article {
 }
+
+/* Credit Card Number */
+.ui-component-credit-card-number .credit-card-images
```
Could we have the opening bracket on the same line as the css selector?

cameronviner commented:
There doesn't seem to be a limit on the amount of characters I can enter on the credit card number field, should there be a limit on the number of characters the user can enter in the field?

manifestinteractive commented:
Addressed comments and merge conflict in Nav.jsx file.  Master has been merged into this PR.

manifestinteractive commented:
__Addressed the following:__

1.  Merged in master and addressed merge conflicts
2.  Updated Unit Tests format to new standard 
3. verified npm test && npm run lint pass

robhuzzey commented on src/components/credit-card-number/code/templates/creditCardNumberTemplate.jsx:
```diff
@@ -0,0 +1,52 @@
+      {images}
+      <div className="credit-card-number">
+        {label}
+        <input
```
Can we make use of the pattern attribute in HTML5 to help out here?: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#attr-pattern

robhuzzey commented on src/components/credit-card-number/code/views/creditCardNumberView.jsx:
```diff
@@ -0,0 +1,105 @@
+    return card;
+  },
+
+  getCard: function(cardNumber) {
```
May be overkill, but there's a way to get card type in a handy little library: https://github.com/Shortbreaks/iinChecker

robhuzzey commented on docs/src/Components.jsx:
```diff
@@ -81,6 +81,29 @@ var Components = React.createClass({
 
+        <article>
+          <h3 id="credit_card_number">Credit Card Number</h3>
+          <p>Masked Credit Card Number input</p>
```
The number isn't "Masked" as I expected... is this a correct description? I would have expected to see it mask as I type (like a password field) based on this description.

jodiedoubleday replied:
Tons of usability studies say don't mask, same thing with passwords now

robhuzzey commented:
Looks good to me... few comments, nothing major.

cameronviner commented:
The cards images seem to have got bigger: https://cloudup.com/cI4LVG1tHCo and entering a card number no longer greys out/colours in the appropriate card.

Also if you try to amend any number apart form the last two it jumps to the end of the input. To recreate this enter the VISA number and then try and delete 85.

manifestinteractive commented:
This is ready for review. Forgot to update CSS class names after removing the ui- prefix ( as requested ).  Added selection range to address input cursor jumping to end if editing text within input.  Also merged in latest master.

jodiedoubleday commented:
There is still a slight issue with the credit card field.
When you type in a credit card number (rather than copy and paste) the formatting goes AWOL
Screencast here - http://quick.as/wmwyfg25q

I hope the screencast makes sense

manifestinteractive commented:
Swapped out custom Validation for third-party Payment library.  Also added error message but made sure it was a property that can be passed in, or set to empty, if desired.

robhuzzey commented on src/components/credit-card-number/code/views/creditCardNumberView.jsx:
```diff
@@ -0,0 +1,78 @@
+    };
+  },
+
+  handleChange: function(elm) {
```
Being pedantic here but the argument ```elm``` here makes me thing you are passing an element into the handle change method when in fact it's an event (convention is that it's ```e```)

robhuzzey commented:
Only one comment about a variable name but apart from that... +1 from me.

cameronviner commented:
Looks good to me, :+1:

manifestinteractive commented:
@cameronviner @robhuzzey @jodiedoubleday Can someone press that fancy green "Merge pull request" button?

lukehansell-hx commented on src/components/credit-card-number/code/views/creditCardNumberView.jsx:
```diff
@@ -0,0 +1,78 @@
+      placeHolder: '• • • •   • • • •   • • • •   • • • •',
+      showImages: false,
+      errorMessage: 'Invalid Card Number'
+    };
```
no default is supplied for defaultValue and not providing one on the element declaration displays the invalid error for some reason.

lukehansell-hx commented on src/components/credit-card-number/scripts/build-dev.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
```
can be removed

manifestinteractive commented:
@lukehansell-hx I updated this component using the other credit card date feedback as well.  Has better unit tests and addresses the other comments listed.
